src/Compilers/CSharp/Portable/Symbols/TypeUnification.cs

```diff
@@ -155,18 +153,15 @@ private static bool CanUnifyHelper(TypeWithAnnotations t1, TypeWithAnnotations t
 
                         NamedTypeSymbol nt1 = (NamedTypeSymbol)t1.Type;
                         NamedTypeSymbol nt2 = (NamedTypeSymbol)t2.Type;
-                        if (!nt1.IsGenericType)
-                        {
-                            return !nt2.IsGenericType && TypeSymbol.Equals(nt1, nt2, TypeCompareKind.ConsiderEverything2);
-                        }
-                        else if (!nt2.IsGenericType)
+                        if (!nt1.IsGenericType || !nt2.IsGenericType)
                         {
+                            Debug.Assert(!TypeSymbol.Equals(nt1, nt2, TypeCompareKind.IgnoreNativeIntegers));
```

Comment:
>Debug.Assert(!TypeSymbol.Equals(nt1, nt2, TypeCompareKind.IgnoreNativeIntegers)); [](start = 28, length = 81)

I think it would be good to add a comment why do we assume this would be the case. Since the logic with previous comparisons got extracted into a helper, it is not obvious.